File: apps/android/app/src/main/kotlin/org/help313/app/MapPalette.kt
// The map's colours, light and dark, exactly the tokens the web map draws with (apps/web/src/style.css as of
// 2026-09-21), including the contrast fixes of docs/ACCESSIBILITY-AUDIT-2026-09-20: every street now clears 3:1
// against the land and against a park, the city edge is a stroked line rather than a 1.20:1 change of shade, and
// the ground outside the service area is a hatch whose own lines clear 3:1 rather than a second pale fill.
//
// The values themselves are in res/values/colors.xml and res/values-night/colors.xml, so Android switches theme on
// its own and nothing here reads a preference. **MapView.kt never names a colour value**: the layer rules in
// MapLayers.kt carry a token name ("bus", "gwOpen"), and this file is the only place a name becomes a number. That
// is what makes the "subway" style (docs/MAP-STYLE.md) a table of tokens rather than an edit to the painter: its
// `--tr-*` tokens and the quietened basemap are NUMBERS in MapStyle.kt (where `:core` holds them to style.css, token
// for token, and computes the spec's contrast tables), and this file only picks the light or dark, plain or
// high-contrast value.
//
// Colour never carries a meaning on its own anywhere on this map: a greenway phase also has its own dash pattern
// and its own words, a status is words, and every layer is named in the switcher, in the tapped card and in the
// list (Differentiate Without Color).
package org.help313.app

import android.content.Context

class MapPalette(private val context: Context) {

    private fun c(id: Int) = UI.color(context, id)

    /** Light or dark, from the same configuration the resources above were resolved with. */
    val scheme: MapScheme =
        if ((context.resources.configuration.uiMode and android.content.res.Configuration.UI_MODE_NIGHT_MASK) ==
            android.content.res.Configuration.UI_MODE_NIGHT_YES
        ) MapScheme.DARK else MapScheme.LIGHT

    /**
     * Whether this person asked the phone for more contrast. **`UiModeManager.getContrast()`, Android 14 and up** —
     * the one public, reflection-free answer the platform gives. `AccessibilityManager.isHighTextContrastEnabled` is
     * hidden until API 36 and this app does not reach for hidden methods, so on Android 7 to 13 the answer is "no":
     * the subway tones are then the plain ones, which already clear 3:1 everywhere (MapStyleTest), and the basemap
     * may be quietened. With it on: the high-contrast tones, casing + 1, rings + 0.5, and never a quiet basemap.
     */
    val highContrast: Boolean = try {
        android.os.Build.VERSION.SDK_INT >= 34 &&
            (context.getSystemService(android.app.UiModeManager::class.java)?.contrast ?: 0f) > 0f
    } catch (_: Throwable) {
        false
    }

    /** A `--tr-*` token of the subway style, as a colour for this theme and this contrast setting. */
    fun transit(token: TransitToken): Int = TransitPalette.color(token, scheme, highContrast).argb

    /** The words on a badge. */
    val badgeText: Int get() = TransitPalette.badgeText(scheme).argb

    /** The quietened basemap's six tokens (subway only, and never with high contrast: `basemapTokens`). */
    val quiet: QuietBasemap = QuietBasemap.tokens(scheme)

    // The ground
    val outside = c(R.color.map_out)
    val outsideInk = c(R.color.map_out_ink)
    val land = c(R.color.map_land)
    val park = c(R.color.map_park)
    val parkInk = c(R.color.map_park_ink)

    // The streets
    val road = c(R.color.map_road)
    val main = c(R.color.map_main)
    val freeway = c(R.color.map_fwy)
    val ink = c(R.color.map_ink)

    // Everything that sits on top
    val surface = c(R.color.surface)
    val greenwayCase = c(R.color.gw_case)
    val me = c(R.color.map_me)
    val brand = c(R.color.brand)

    /** The wash on the one area outline a tap picked out. It carries no number (docs/13, rule 1). */
    val brandSoft = c(R.color.brand_soft)

    /**
     * Every neighbourhood and city boundary, in one token (docs/MAP-STYLE.md section 15.2) — so a boundary is one
     * thing wherever it is drawn, in either map style. The high-contrast value is picked here, as the subway
     * tones are, rather than by the painter.
     */
    val boundary = c(if (highContrast) R.color.map_bnd_more else R.color.map_bnd)

    /**
     * The colour a layer style or a greenway phase names (MapLayers.kt), resolved here so that the rules stay free
     * of Android and `:core` can run them on a plain JDK.
     */
    fun named(key: String): Int = when (key) {
        "bus" -> c(R.color.lyr_bus)
        "smart" -> c(R.color.lyr_smart)
        "rail" -> c(R.color.lyr_rail)
        "bike" -> c(R.color.lyr_bike)
        "gwOpen" -> c(R.color.gw_open)
        "gwBuild" -> c(R.color.gw_build)
        "gwFund" -> c(R.color.gw_fund)
        "gwPlan" -> c(R.color.gw_plan)
        // The two tones one trip is drawn in (DirWords.rideToken). A ride names its agency's own layer token
        // above; "routeRide" is what an agency we carry no layer for falls back to.
        "routeWalk" -> c(R.color.route_walk)
        "routeRide" -> c(R.color.route_ride)
        else -> brand
    }

    /** The dot colour for one of our own listing groups (MAP_GROUPS). */
    fun group(id: String): Int = when (id) {
        "food" -> c(R.color.grp_food)
        "shelter" -> c(R.color.grp_shelter)
        "health" -> c(R.color.grp_health)
        "rec" -> c(R.color.grp_rec)
        "work" -> c(R.color.grp_work)
        "kids" -> c(R.color.grp_kids)
        "things" -> c(R.color.grp_things)
        "paperwork" -> c(R.color.grp_paperwork)
        else -> brand
    }
}

File: apps/ios/HelpApp/MapPalette.swift
// The map's colours. The `standard` basemap, the greenway and the layer colours are the web's own tokens
// (apps/web/src/style.css `--map-*`, `--gw-*`, `--lyr-*`) in all four of their values — light, dark, and each
// with Increase Contrast, the web's `prefers-contrast: more` — and the numbers live in HelpCore
// (`StandardPalette`), where `swift test` holds them to the web's file token for token and to 3:1 for every pair
// a frame puts side by side: every street against the land AND a park, the hatch outside the four cities against
// both grounds, every line against its casing (docs/ACCESSIBILITY-AUDIT-2026-09-20, WCAG 1.4.11).
//
// Colour never carries a meaning on its own anywhere on this map: a greenway phase also has its own dash pattern
// and its own words, a status is words, and every layer is named in the switcher, in the tapped card and in the
// list (Differentiate Without Color).
import HelpCore
import SwiftUI
import UIKit

private func mapPair(_ light: UInt32, _ dark: UInt32) -> Color {
    func c(_ v: UInt32) -> UIColor {
        UIColor(red: CGFloat((v >> 16) & 255) / 255, green: CGFloat((v >> 8) & 255) / 255, blue: CGFloat(v & 255) / 255, alpha: 1)
    }
    return Color(uiColor: UIColor { $0.userInterfaceStyle == .dark ? c(dark) : c(light) })
}

extension Color {
    /// A colour from HelpCore's palette tables (the `--tr-*` tokens and the quietened basemap of docs/MAP-STYLE.md),
    /// where the numbers live so that `swift test` can hold them to the spec's contrast ratios. They arrive
    /// already resolved for light or dark and for Increase Contrast, so this is a plain colour, not a dynamic one.
    init(rgb c: RGB) { self.init(red: Double(c.r) / 255, green: Double(c.g) / 255, blue: Double(c.b) / 255) }
}

/// One of the web's map tokens as a dynamic colour: it follows light / dark AND Increase Contrast, wherever it is
/// used (the Canvas resolves it against the environment, as it does every other colour).
private func mapToken(_ t: MapToken) -> Color {
    func c(_ v: RGB) -> UIColor { UIColor(red: CGFloat(v.r) / 255, green: CGFloat(v.g) / 255, blue: CGFloat(v.b) / 255, alpha: 1) }
    return Color(uiColor: UIColor { traits in
        c(StandardPalette.color(t, scheme: traits.userInterfaceStyle == .dark ? .dark : .light,
                                increasedContrast: traits.accessibilityContrast == .high))
    })
}

enum MapColor {
    // The ground
    static let outside = mapToken(.out)                     // beyond the four cities, under the hatch
    static let outsideInk = mapToken(.outInk)               // the hatch itself: 3:1 against both grounds
    static let land = mapToken(.land)
    static let park = mapToken(.park)
    static let parkInk = mapToken(.parkInk)
    // The streets: all three clear 3:1 against the land and against a park; width carries the hierarchy
    static let road = mapToken(.road)
    static let main = mapToken(.main)
    static let freeway = mapToken(.freeway)
    static let ink = mapToken(.ink)                         // street names
    // The greenway, drawn like a transit line
    static let gwOpen = mapToken(.gwOpen)
    static let gwBuild = mapToken(.gwBuild)
    static let gwFund = mapToken(.gwFund)
    static let gwPlan = mapToken(.gwPlan)
    static let gwCase = mapToken(.gwCase)                   // the casing under the greenway AND under every transport line
    // The transport layers
    static let bus = mapToken(.bus)
    static let smart = mapToken(.smart)
    static let rail = mapToken(.rail)
    static let bike = mapToken(.bike)
    // Our own listings, one colour per help group
    static let grpFood = mapPair(0xA16207, 0xE3B341)
    static let grpShelter = mapPair(0x4338CA, 0xA5B4FC)
    static let grpHealth = mapPair(0xBE185D, 0xF9A8D4)
    static let grpRec = mapPair(0x15803D, 0x4FCF93)
    static let grpWork = mapPair(0x0E7490, 0x67E8F9)
    static let grpKids = mapPair(0x7E22CE, 0xD8B4FE)
    static let grpThings = mapPair(0x7C2D12, 0xFDBA74)
    static let grpPaperwork = mapPair(0x475569, 0xCBD5E1)
    /// Where this phone is, when the person asked for it.
    static let me = mapPair(0x0A57C2, 0x8AB8FF)
    /// A city or neighbourhood outline (`place:areas`): the web's `--muted` for the dashed line, `--ink` for the
    /// name, `--focus` for the one that was tapped. A boundary is a line and a name and never a shade.
    static let areaLine = mapPair(0x47584F, 0xA3B5AA)
    static let areaInk = mapPair(0x10201A, 0xE9F1EC)
    static let focus = mapPair(0x0A57C2, 0x8AB8FF)
    /// A trip, on a Directions map: the web's `--route-walk` and `--route-ride` (apps/web/src/style.css).
    static let routeWalk = mapPair(0x9D1B6A, 0xFF8AD4)
    static let routeRide = mapPair(0x00558F, 0x73C2FF)

    /// The name a layer style or a greenway phase carries (HelpCore/MapLayers.swift), resolved to a colour here so
    /// that the rules stay free of SwiftUI and `swift test` can reach them.
    static func named(_ key: String) -> Color {
        switch key {
        case "bus": return bus
        case "smart": return smart
        case "rail": return rail
        case "bike": return bike
        case "gwOpen": return gwOpen
        case "gwBuild": return gwBuild
        case "gwFund": return gwFund
        case "gwPlan": return gwPlan
        case routeWalkToken: return routeWalk
        case routeRideToken: return routeRide
        default: return .brand
        }
    }

    /// The dot colour for one of our own listing groups (MAP_GROUPS).
    static func group(_ id: String) -> Color {
        switch id {
        case "food": return grpFood
        case "shelter": return grpShelter
        case "health": return grpHealth
        case "rec": return grpRec
        case "work": return grpWork
        case "kids": return grpKids
        case "things": return grpThings
        case "paperwork": return grpPaperwork
        default: return .brand
        }
    }
}
